File: clojure/dnd-character/src/dnd_character.clj
(ns dnd-character)

(defn score-modifier
  "Calculates the modifier of the given score."
  [score]
  (clojure.math/floor-div (- score 10) 2))

(defn rand-ability
  "Generates a random ability."
  []
  (reduce + (rest (sort (repeatedly 4 #(inc (rand-int 6)))))))

(defrecord DndCharacter
  [strength dexterity charisma wisdom intelligence constitution hitpoints])

(defn rand-character
  "Generates a random character."
  []
  (let [ability-scores (repeatedly 6 rand-ability)
        [strength dexterity charisma wisdom intelligence constitution] ability-scores
        hitpoints (+ 10 (score-modifier constitution))]
    (->DndCharacter strength dexterity charisma wisdom intelligence constitution hitpoints)))

File: elixir/dnd-character/lib/dnd_character.ex
defmodule DndCharacter do
  @type t :: %__MODULE__{
          strength: pos_integer(),
          dexterity: pos_integer(),
          constitution: pos_integer(),
          intelligence: pos_integer(),
          wisdom: pos_integer(),
          charisma: pos_integer(),
          hitpoints: pos_integer()
        }

  defstruct ~w[strength dexterity constitution intelligence wisdom charisma hitpoints]a

  @spec modifier(pos_integer()) :: integer()
  def modifier(score), do: score - 10 |> Integer.floor_div(2)

  @spec ability :: pos_integer()
  def ability, do:
    for(_roll <- 1..4, do: :rand.uniform(6))
    |> Enum.sort()
    |> Enum.drop(1)
    |> Enum.sum()

  @spec character :: t()
  def character do
    constitution = ability()

    %DndCharacter {
      strength: ability(),
      dexterity: ability(),
      constitution: constitution,
      intelligence: ability(),
      wisdom: ability(),
      charisma: ability(),
      hitpoints: 10 + modifier(constitution)
    }
  end
end
